common/transport/amqp/devdoc/sender_link_requirements.md

```diff
@@ -4,73 +4,65 @@
 
-## Example usage
-
-```typescript
-import * as amqp10 from 'amqp10';
-import { AmqpMessage } from './amqp_message';
-
-const linkAddress = 'exampleAddress';
-const amqp10Client = new amqp10.AmqpClient(null);
-
-const senderLink = new SenderLink(linkAddress, null, amqp10Client);
-senderLink.on('errorReceived', (err) => {
-  console.error(err.toString());
-});
-
-senderLink.send(new AmqpMessage(''), linkAddress, (err) => {
-  if (err) {
-    console.error(err.toString());
-  } else {
-    console.log('message successfully sent');
-  }
-});
-```
 
 ## Public Interface
 
-### constructor(linkAddress: string, linkOptions: any, amqp10Client: amqp10.AmqpClient)
+### constructor(linkAddress: string, linkOptions: SenderOptions, session: Session)
 
 **SRS_NODE_AMQP_SENDER_LINK_16_002: [** The `SenderLink` class shall inherit from `EventEmitter`. **]**
 
 **SRS_NODE_AMQP_SENDER_LINK_16_003: [** The `SenderLink` class shall implement the `AmqpLink` interface. **]**
 
+**SRS_NODE_AMQP_SENDER_LINK_16_001: [** The `SenderLink` internal state machine shall be initialized in the `detached` state. **]**
 ### attach(callback: (err?: Error) => void): void
 
-**SRS_NODE_AMQP_SENDER_LINK_16_004: [** The `attach` method shall use the stored instance of the `amqp10.AmqpClient` object to attach a new link object with the `linkAddress` and `linkOptions` provided when creating the `SenderLink` instance. **]**
+
+**SRS_NODE_AMQP_SENDER_LINK_16_004: [** The `attach` method shall use the stored instance of the `rhea` session object to attach a new link object with the combined `linkAddress` and `linkOptions` provided when creating the `SenderLink` instance. **]**
 
 **SRS_NODE_AMQP_SENDER_LINK_16_022: [** The `attach` method shall call the `callback` if the link was successfully attached. **]**
 
-**SRS_NODE_AMQP_SENDER_LINK_16_006: [** The `SenderLink` object should subscribe to the `detached` event of the newly created `amqp10` link object. **]**
+**SRS_NODE_AMQP_SENDER_LINK_06_007: [** The `attach` method shall immediately invoke the `callback` if already in an attached state. **]**
 
-**SRS_NODE_AMQP_SENDER_LINK_16_007: [** The `SenderLink` object should subscribe to the `errorReceived` event of the newly created `amqp10` link object. **]**
+**SRS_NODE_AMQP_SENDER_LINK_06_001: [** If the `detach` method is invoked on the `SenderLink` while still attaching, the SenderLink shall detach.  It will indicate the error to the callback for the `detach` as well as the callback to the `attach` **]**
 
-**SRS_NODE_AMQP_SENDER_LINK_16_008: [** If the `amqp10.AmqpClient` fails to create the link the `callback` function shall be called with this error object. **]**
+**SRS_NODE_AMQP_SENDER_LINK_06_003: [** If the `forceDetach` method is invoked on the `SenderLink` while still attaching, the SenderLink shall detach.  It will indicate the error to the callback to the `attach` and emit an `error` event on the `SenderLink`.**]**
```

Comment:
Both do.  Changed requirments for sender and recevier to read:  With the error supplied to the forceDetach, the `attach` callback will also be invoked.  If the error is NOT falsy it will also be emitted as the argument to the `error` event.] */
        it('forceDetach while attaching invokes the attach callback with an error (if provided) and emits an `error` event with the (potentially) supplied error.

---
In reply to: [209125785](https://github.com/Azure/azure-iot-sdk-node/pull/344#discussion_r209125785) [](ancestors = 209125785)